[documentation] deepsparse user guide (#959)
neuralmagic/deepsparse: user-guide/deepsparse-server.md

# DeepSparse Server

DeepSparse Server wraps [Pipelines](deepsparse-pipelines.md) with a REST API, making it easy to stand up a inference 
serving endpoint running DeepSparse.

## Quickstart

DeepSparse Server is launched from the CLI. Just like DeepSparse Pipelines, all we 
have to do is pass a task and a model.

Spin up sentiment analysis endpoint with a 90% pruned-quantized BERT model:
```bash
deepsparse.server \
  --task sentiment-analysis \
  --model_path zoo:nlp/sentiment_analysis/obert-base/pytorch/huggingface/sst2/pruned90_quant-none 
```

In this case, we used a SparseZoo stub, which instructs the Server to download the relevant
ONNX file from the SparseZoo. To deploy your own model, pass a path to a `model.onnx` file or a 
folder containing the  `model.onnx` and supporting files (e.g., the Hugging Face `tokenizer.json` and `config.json`).

> Let's make a request over HTTP. Since the Server is a wrapper around Pipelines, 

Reviewer: ```suggestion
Make a request over HTTP; since the Server is a wrapper around Pipelines, 
```

> we can send raw data to the endpoint and recieve the post-processed predictions:

Reviewer: ```suggestion
we can send raw data to the endpoint and receive the post-processed predictions:
```

```

For full usage, run:
```bash
deepsparse.server --help
```

## Supported Use Cases

DeepSparse Server supports all tasks available in DeepSparse Pipelines. Check out the use case guides for more details on task-specific APIs.

**Computer Vision**:
- Image Classification: `task="image_classification"`
- Object Detection: `task="yolo"`
- Instance Segmentation: `task="yolact"`

**Natural Language Processing**:
- Embedding Extraction: `task="transformers_embedding_extraction"`
- Text Classification: `task="text-classification"`
- Zero Shot Text Classification: `task="zero-shot-text-classification"` 
- Sentiment Analysis: `task="sentiment-analysis"`
- Token Classification: `task="token-classification"`
- Question Answering: `task="question-answering"`

## Swagger UI

FastAPI's Swagger UI enables you to view your Server's routes and to make sample requests. Navigate to the `/docs` 
route (e.g., `http://localhost:5543/docs`) to try it out.

<p align="center">
  <img src="https://raw.githubusercontent.com/neuralmagic/deepsparse/main/src/deepsparse/server/img/swagger_ui.png"/>
</p>

## Server Configuration

You can configure DeepSparse Server via YAML files.

### Basic Example

> Let's walk through a basic example deploying via a configuration file.

Reviewer: ```suggestion
Let us walk through a basic example of deploying via a configuration file.
```

```

We can then spin up with the `--config-file` argument:

```bash
deepsparse.server \
  --config-file config.yaml
```

Sending a request:
```python
import requests
url = "http://localhost:5543/predict"
obj = {"sequences": "I love querying DeepSparse launched from a config file!"}
print(requests.post(url, json=obj).text)

# >>> {"labels":["positive"],"scores":[0.9136188626289368]}
```

### Server Level Options

At the server level, there are a few arguments that can be toggled.

#### Physical Resources
`num_cores` specifies the number of cores that DeepSparse runs on. By default,
DeepSparse runs on all available cores.

#### Scheduler
`num_workers` configures DeepSparse's scheduler. 

If `num_workers = 1` (the default), DeepSparse uses its "synchronous" scheduler, which allocates as many resources as possible 
to each request. This format is optimizes per-request latency. By setting `num_workers > 1`, DeepSparse 
utilizes its multi-stream scheduler, which processes multiple requests at the same time. 
In deployment scenarios with low batch sizes and high core counts, using the "multi-stream" scheduler 
can increase throughput by allowing DeepSparse to better saturate the cores.

> The following configuration creates a Server with DeepSparse running on 2 cores, with 2 input streams,

Reviewer: ```suggestion
The following configuration creates a Server with DeepSparse running on two cores, with two input streams,
```

dense unoptimized BERT and one using a 90% pruned-quantized BERT.

```yaml
# multiple-endpoint-config.yaml
endpoints:
  - task: sentiment-analysis
    model: zoo:nlp/sentiment_analysis/obert-base/pytorch/huggingface/sst2/pruned90_quant-none
    route: /sparse/predict
    name: sparse-sentiment-analysis

  - task: sentiment-analysis
    model: zoo:nlp/sentiment_analysis/obert-base/pytorch/huggingface/sst2/base-none
    route: /dense/predict
    name: dense-sentiment-analysis
```

Spinning up:
```bash
deepsparse.server \ 
  --config-file multiple-endpoint-config.yaml
```

Making a request:
```python
import requests

obj = {"sequences": "I love querying the multi-model server!"}

sparse_url = "http://localhost:5543/sparse/predict"
print(f"From the sparse model: {requests.post(sparse_url, json=obj).text}")

dense_url = "http://localhost:5543/dense/predict"
print(f"From the dense model: {requests.post(dense_url, json=obj).text}")

# >>> From the sparse model: {"labels":["positive"],"scores":[0.9942120313644409]}
# >>> From the dense model: {"labels":["positive"],"scores":[0.998753547668457]}
```

### Endpoint Level Configuration

> We can also configure properties of each endpoint, including task specific

Reviewer: ```suggestion
We can also configure the properties of each endpoint, including task-specific
```

    would not continue watching. Roddenberry's ashes must be turning in their orbit as this dull, cheap, poorly edited (watching it without advert breaks \
    really brings this home) trudging Trabant of a show lumbers into space. Spoiler. So, kill off a main character. And then bring him back as another actor. \
    Jeeez! Dallas all over again."}

# batch 3
short_obj = {"sequences": [
  "I love how easy it is to configure DeepSparse Server!",
  "It was very challenging to configure my old deep learning inference platform",
  "YAML is the best format for configuring my infrastructure"
]}

document_classification_url = "http://localhost:5543/text-classification/predict"
print(requests.post(document_classification_url, json=document_obj).text)

sentiment_analysis_url = "http://localhost:5543/sentiment-analysis/predict"
print(requests.post(sentiment_analysis_url, json=short_obj).text)

# >>> {"labels":[["0","1"]],"scores":[[0.9994900226593018,0.0005100301350466907]]}
# >>> {"labels":["positive","negative","positive"],"scores":[0.9665533900260925,0.9952980279922485,0.9939143061637878]}
```

> Checkout the use case pages for detailed documentation on task-specific arguments that can be applied to the Server via `kwargs`.

Reviewer: ```suggestion
Check out the use case pages for detailed documentation on task-specific arguments that can be applied to the Server via `kwargs`.
```

## Custom Use Cases

> Stay tuned for documentation on Using a custom DeepSparse Pipeline wihin the Server!

Reviewer: ```suggestion
Stay tuned for documentation on using a custom DeepSparse Pipeline within the Server!
```
